### ISSUE TITLE
Clarify thread safety behavior of Dogstatsd 5.x

### ISSUE THREAD

**Issue by petergoldstein** [DataDog/dogstatsd-ruby#178]
I'd like to get some clarity (ideally published in the README) around thread safety for Dogstatsd 5.x and going forward.

This is important because:

1. There are implementations in a number of significant libraries that seem to assume it's thread safe, and I don't believe the underlying socket is actually thread safe (e.g. [racecar](https://github.com/zendesk/racecar/blob/d4c156e55f8fc00984405bff81b2c77a27b14775/lib/racecar/datadog.rb) and [ruby-kafka](https://github.com/zendesk/ruby-kafka/blob/a56d16bf1ff30d784c4617159033569045e157eb/lib/kafka/datadog.rb)
2. The asynchronous flush seems to have substantially changed the socket behavior - clients need to be explicitly closed, or there will be a file handle leak

Re: 1, I'd like to know if I should open up issues on those libraries to update to a thread safe usage.

Re: 2, in our own application we had historically been using a Thread.context to store a Dogstatsd client per thread, to ensure thread safety, and this was not an issue.  As of 5.x, because we're not calling close explicitly, we have a file handle leak.  If the client is thread safe, then we can just reduce all of this to a single Dogstatsd client.  But looking at the implementation, I suspect it isn't, and a connection pool will be required.  We'd like to get clarity on the best path forward.

**Comment by kbogtob** [comment]
Hello @petergoldstein! I was the developer rewriting the new threading model. I've just added some documentation about it in PR #183.
TL;DR: We now have a background running thread writing data to the agent's socket. There is no more IO when using the statsd client.

To answer your questions:

1. We designed it so that the Statsd client can be shared between threads and the background thread will do all the critical work by itself. In your example, using a same statsd client, all your threads will be able to message the sender thread in background.

2. You're right about the changed behavior of the client. If you don't close it, you are not sure about your data reaching the agent. We are adding PRs #180 and #181 to add some "flush on client close logic" and a new auto-close on program exit feature so that no data loss is happening on short scripts and such.

**Comment by petergoldstein** [comment]
That's super helpful @kbogtob.  I'll wait for the additions in PR #180 and PR#181, but my basic takeaway is that we should remove all of our thread context code, and we can replace with a single instance, rather than say a connection pool.

@mperham may also be interested in that, as I'm pretty sure he's using a connection pool for Dogstatsd clients in Sidekiq.

**Comment by mperham** [comment]
Moving to a singleton model wouldn’t be my recommendation, that lock becomes a process-wide source of contention. Sidekiq’s dogstatsd support uses a pool of clients so there’s no concurrent usage to worry about. That doesn’t mean you shouldn’t make the client threadsafe to ensure safety everywhere but providing a small pool vs a singleton is still a good idea to minimize contention.

**Comment by petergoldstein** [comment]
@mperham Fair point.

Frankly I'm less concerned about latency and throughput than I am about data corruption and predictable behavior.  Thread safety gets that for me, once the flush/close behavior is updated.  So that's the big win for us.

I'm also not sure how expensive that lock contention is in practice.  That said, provided that the individual Dogstatsd clients aren't too heavy, there's not really a downside to a pool - especially when you're doing a synchronous flush of stats on a regular basis.  We'll have to see.

**Comment by remeh** [comment]
Hey, since we completed [information on the threading model here in the README](https://github.com/DataDog/dogstatsd-ruby#threading-model), plus, we also documented [the pitfall of v5.x](https://github.com/DataDog/dogstatsd-ruby#migrating-from-v4x-to-v5x) (i.e. close needed to not leak threads when multiple instances of the agent are spawned in an app lifecycle), I'm thinking about closing that issue. What do you think?

**Comment by petergoldstein** [comment]
I'm ok with closing
